core/src/test/scala/unit/kafka/log/LogTest.scala

```diff
@@ -1013,6 +1013,54 @@ class LogTest extends JUnitSuite {
   }
 
   @Test
+  def testOverCompactedLogRecovery(): Unit = {
+    // append some messages to create some segments
+    val logProps = new Properties()
+    logProps.put(LogConfig.SegmentBytesProp, 1000: java.lang.Integer)
+    logProps.put(LogConfig.MaxMessageBytesProp, 64*1024: java.lang.Integer)
+    logProps.put(LogConfig.IndexIntervalBytesProp, 1: java.lang.Integer)
+    val config = LogConfig(logProps)
+    val log = new Log(logDir,
+      config,
+      recoveryPoint = 0L,
+      time.scheduler,
+      time)
+    val set1 = new ByteBufferMessageSet(
+      compressionCodec = NoCompressionCodec,
+      offsetSeq = Seq(0),
+      messages = new Message("v1".getBytes(), "k1".getBytes(), Message.NoTimestamp, magicValue = Message.CurrentMagicValue))
+    val set2 = new ByteBufferMessageSet(
```

Comment:
To get around this, we could roll a segment with a base offset of the first offset in the to be appended message set, instead of log end offset. However, it's not easy to get the first offset in the follower w/o decompression since the offset of the wrapper message is always the offset of the last message in the set. KIP-98 proposes a new message format that allows us to get the first offset w/o decompression. Not sure if there are other workarounds in the interim.